### ISSUE TITLE
Flutter does not call detachFromEngine when the shell is reset as part of applicationWillTerminate

### ISSUE THREAD

**Issue by fuzzybinary** [flutter/flutter#126671]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Create an iOS plugin that sends messages back to Flutter on a method channel.
2. Send messages to the method channel frequently from a separate thread on the application.
3. Register your plugin with the registrar using `-[FlutterPluginRegistrar publish:]` and implementing `-[FlutterPlugin detachFromEngineForRegistrar:]` to stop the message send.
4. Shutdown the app.

### Expected results

Flutter calls `-[FlutterPlugin detachFromEngineForRegistrar]` and the app does not crash.

### Actual results

In cerain situations, the app will crash with an assertion: "Sending a message before the FlutterEngine has been run." which comes from this line: https://github.com/flutter/engine/blob/8a95a0ea2699aab2d46cbd76fe8fbed551abb844/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm#L1192

### Code sample

<details><summary>Code sample</summary>

```dart
[Paste your code here]
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[!] Flutter (Channel stable, 3.7.6, on macOS 13.3.1 22E772610a darwin-arm64, locale en-US)
    ! Warning: `dart` on your path resolves to /opt/homebrew/Cellar/dart/2.18.7/libexec/bin/dart, which is not inside your current Flutter SDK checkout at /Users/jeff.ward/Projects/flutter.
      Consider adding /Users/jeff.ward/Projects/flutter/bin to the front of your path.
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.1)
[✓] VS Code (version 1.78.1)
[✓] Connected device (6 available)
[✓] HTTP Host Availability

! Doctor found issues in 1 category.
```

</details>

**Comment by fuzzybinary** [comment]
Very tough to provide a code sample here, but I'll come back to it and provide a github repo if I can.

Some more information - we're suffering from this crash in the Datadog plugin, see https://github.com/DataDog/dd-sdk-flutter/issues/341 (this was closed but the issue persists in newer versions).  I've tracked this down to the following code path:

1. iOS sends an `ApplicationWillTerminate` notification
2. `-[FlutterViewController applicationWillTerminate:]` is called followed by
3. `-[FlutterEngine destroyContext]` which shuts down the platform channel via `_shell.reset()`

I feel like `destroyContext` should either:
1. Detach plugins from the engine, similar to this codepath in `-[FlutterEngine dalloc]` (https://github.com/flutter/engine/blob/8a95a0ea2699aab2d46cbd76fe8fbed551abb844/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm#L256)
2. Have a separate notification for plugins to inform them that the shell is being reset

**Comment by darshankawar** [comment]
Thanks for the detailed report @fuzzybinary 
Would also be good to provide us a demo repo or a code sample that we can use to verify this behavior further.

**Comment by fuzzybinary** [comment]
Hi @darshankawar,

Here's my example: https://github.com/fuzzybinary/flutter_shutdown_crash

In my opinion, this should not crash on a force quit.

**Comment by stuartmorgan** [comment]
> 4\. Shutdown the app.

Meaning swiping it away in the app switcher UI, I assume (given the stack)?

**Comment by fuzzybinary** [comment]
Yes. Be forewarned that taking too much time debugging when quitting like this will terminate the debug session without going through all the shutdown code (like the FlutterEngine dealloc) which makes this incredibly hard to debug.

**Comment by stuartmorgan** [comment]
> In my opinion, this should not crash on a force quit.

It definitely shouldn't crash.

That said, from a prioritization standpoint this is a pretty minor issue; flows that call `applicationWillTerminate` on iOS are quite rare in real-world usage (most iOS users don't even know you can do that), and the effect isn't visible to non-developer users because the process going away was the expected outcome.

(If we eventually combine more iOS and macOS engine codepaths, this would be much more important.)

**Comment by fuzzybinary** [comment]
Not going to disagree, but unfortunately it does show up in crash trackers and gets reported back to library maintainers 🙂.  I think that possibly (though I can't be sure) memory pressure can also cause this exit code path. Again, deosn't affect users but does get reported to crash trackers.

Given direction I could potentially take this on. I feel like the correct answer is to move this code path (https://github.com/flutter/engine/blob/8a95a0ea2699aab2d46cbd76fe8fbed551abb844/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm#L256) into `-[FlutterEngine destroyContext]`, but don't want to spend too much time on it if that's not the right answer.

**Comment by stuartmorgan** [comment]
> I think that possibly (though I can't be sure) memory pressure can also cause this exit code path.

It would be extremely surprising if, in the face of memory pressure, iOS chose to **activate background applications** in order to let them run arbitrary code, rather than just terminate them silently in the background as applications are always supposed to be designed to assume can happen.

In many years of iOS development prior to Flutter, the only way I ever remember seeing this codepath was via the user manually swiping away the **currently active** application from the switcher.

**Comment by fuzzybinary** [comment]
Yes, I think you're right, that makes sense. The only way I've seen this codepath trigger is through swiping away in the application in the switcher. Though, I have seen it trigger from the *inactive* application in the switcher.

**Comment by percula** [comment]
Since migrating to `FlutterSceneDelegate` (3.41.2) this became my app's top production crash (~97 events/80 users a week). It didn't exist in the release before the migration.

The problem is that `sceneWillDisconnect:` now calls `appOrSceneWillTerminate` → `[engine destroyContext]`, and iOS disconnects backgrounded scenes routinely (in my crash reports, 6-51 minutes after backgrounding). So the shell gets destroyed regularly, plugins never get `detachFromEngineForRegistrar:`, and the next native callback into a stale `eventSink` hits the `NSAssert` in `sendOnChannel:`. For me that's mostly connectivity_plus's `NWPathMonitor`, with smaller buckets from cloud_firestore and firebase_auth stream handlers. The plugins do implement detach correctly, it just never runs on this path.

I worked around the connectivity case by cancelling my subscription on `AppLifecycleState.paused`, but that doesn't help with streams plugins own internally (Firestore snapshot listeners etc.), and the `detached` message is sent synchronously right before the shell is destroyed so Dart can't clean up at teardown time.

Could `destroyContext` call `detachFromEngineForRegistrar:` before resetting the shell? Happy to share full stacks or test a patch.

**Comment by Wackymax** [comment]
Hi, I managed to reproduce this with a small Flutter app without Firebase or any third-party plugins. I've put the [code, steps and logs in a gist](https://gist.github.com/Wackymax/d0c8fe508991ed5305c1f43264ef1f26).

The app starts a second engine and subscribes to an event channel. The native plugin is published and implements `detachFromEngine(for:)`. If I call `destroyContext()` but keep the engine reference in the AppDelegate, the detach callback never runs. Calling the saved event sink 100 ms later on the main queue then crashes with:

```text
NSInternalInconsistencyException: Sending a message before the FlutterEngine has been run.
```

If I release the engine reference after `destroyContext()`, the detach callback does run, clears the sink, and the app stays running. The plugin itself stays alive for the delayed callback in both cases.

I tested this with Flutter 3.47.2 stable, Dart 3.13.2, Xcode 27 and an iOS 26.5 simulator. This reproduces the engine reset case directly; I haven't reproduced the full scene-disconnection or application-termination flow with this example.

I came across this while working on firebase/flutterfire#18658. Cleaning up the Storage listener helps when the detach callback runs, but it can't handle this case if the engine doesn't notify the plugin before resetting.
